### PR TITLE
storage: clear Entry before reuse

### PR DESCRIPTION
**Pull request by tbg**
See:

https://github.com/cockroachdb/cockroach/issues/18324#issuecomment-328364081

Touches #18324.

### CODE REVIEW

File: pkg/storage/store.go [hunk 1]
```diff
@@ -3429,6 +3429,7 @@ func sendSnapshot(
 	{
 		var ent raftpb.Entry
 		for i := range logEntries {
+			ent.Reset()
```
Comment (petermattis) [inline pkg/storage/store.go]: `ent.Unmarshal` resets internally. Doesn't it?

Comment (tschottdorf) [reply pkg/storage/store.go]: No, see the discussion linked to in this PR. Unless I really messed up both the unit test there and my code inspection.

**Comment by RaduBerinde** [comment]
LGTM

File: pkg/storage/store.go (hunk 1 above)
Comment (petermattis) [reply pkg/storage/store.go]: Oh, it doesn't! That's surprising. Is that a bug in gogoproto? I'm pretty sure `proto.Unmarshal` clears the destination.

**Review by petermattis — APPROVED** [review]
LGTM

**Comment by tschottdorf** [comment]
Yes, `proto.Unmarshal` does and that's also what I looked at first. I'm not sure this is a bug. I'll definitely file something to investigate. This is likely an issue elsewhere.

**Comment by petermattis** [comment]
`git grep '\.Unmarshal(' | grep -v _test.go | grep -v 'proto.Unmarshal(' | grep -v "Binary file" | grep -v 'jsonpb.Unmarshal(' | grep -v '^cmd/'` reveals 29 calls to `\.Unmarshal(`. A quick inspection doesn't show any of them to be problematic because they are all unmarshaling on pristine variables. Of course, I could have missed something. 

Pretty sure this is a bug in gogoproto. From the `proto.Unmarshal` docs:

```
// Unmarshal resets pb before starting to unmarshal, so any
// existing data in pb is always removed. Use UnmarshalMerge
// to preserve and append to existing data.
```

**Comment by tschottdorf** [comment]
Yep, that's the  comment that led me to believe everything was fine until I wrote the unit test. Filed https://github.com/cockroachdb/cockroach/issues/18407 for you (which links to an upstream issue).

**Pull request merged**